### ISSUE TITLE
Support 3 way merge for git conflicts

### ISSUE THREAD

**Issue by rxdeveloper** [microsoft/vscode#37350]
IntelliJ, and all other IntelliJ-based IDE such as WebStorm, support a git merge's conflict resolution tool that it shows the conflict in three columns:
- the first column for Local Changes
- the third column for Changes from Server
- the second column for Conflict Resolution Result

Please see my attached file.
![merge-conflict-resolution](https://user-images.githubusercontent.com/1470309/32250860-c677e4ce-bec0-11e7-82b5-0196d981cc28.png)

I find that it's super easy for us to resolve conflict, because I can compare the Local Changes and the Server Changes on the same windows, and I can see the Resolution Result immediately.

Can we please support that?

**Comment by Auwalms** [comment]
This will be a super cool feature, I use `VSCode` for my day job but often resort to `WebStorm` when I need to resolve a merge conflict due to its interactivity and ease to use.

**Comment by Nodragem** [comment]
Any news on this? this feature would be amazing. On one hand, there is a diff tool (i.e. comparing two files) in VS code, but you cannot edit / chose what to keep; on the other hand, there is a merge conflict tool that allows to chose what to keep (current change / their change). Now, the best would be to have both tools into one tool, such as the tool suggested by @uyhung.

I also wanted to add, for any contributor who comes by: VS Code is an amazing piece of software, keep it up :) !

**Comment by uvaldez** [comment]
That will be an amazing feature... any news on it? does anyone recommend some extension that does something similar?

**Comment by krzysztofwolski** [comment]
I'm using that on PyCharm. Such amazing feature, it helps a lot with avoiding bugs on bigger merges. All linters are available during process, so errors/style violations/unused variables are easy to spot.

**Comment by millanbrankovic** [comment]
Please hurry up with this feature if you're planning to implement it :)

**Comment by igocooper** [comment]
Thumb up 👍 for this one. That's the only thing holding my back to totally switch to VSCode. Would be super helpful and handy.

**Comment by VladimirTyalo** [comment]
The day you implement this feature I will stop using webstorm.

**Comment by argordmel** [comment]
Not only to resolve conflicts, this design should be available to see the history file too.

**Comment by OJFord** [comment]
`git config merge.conflictStyle diff3` and you have this today but with VSCode's in-editor layout - which is vastly better than popping up a dialogue and a new window to view the file you already had open IMO!

**Comment by VladislavAnkudinov** [comment]
Any news? Is it scheduled maybe, or planned to be scheduled?

**Comment by menvil** [comment]
Yes really needed

**Comment by ltsharma** [comment]
this is much needed feature, im using inteliij idea just to resolve conflicts.

**Comment by TimoWestland** [comment]
+1, this would be awesome!

**Comment by yannickbuntsma** [comment]
+1 here! Trying to make the switch, but IMO this is one of the major features missing to switch from WebStorm to VSCode

**Comment by Zielak** [comment]
Can we please **stop** with the +1 comments? It brings nothing to the conversation.
Just watch this thread and wait...

**Comment by stephanmullerNL** [comment]
@Zielak +1 let's stop adding more comments

**Comment by rajjejosefsson** [comment]
Anyone know if there is any plugin that can make this possible ?

**Comment by OJFord** [comment]
@rajjejosefsson As I said [above](https://github.com/Microsoft/vscode/issues/37350#issuecomment-391002625), you can set: (`git config merge.conflictStyle diff3`)
```ini
[merge]
  conflictStyle = diff3
```

to see them/common/us. It won't change the VSCode format/theme, but you will get those three parts to your diff instead of just them/us as is the (git) default.

Unless there's something else (other than opening multiple dialogues just to get back to view a diff in the file you already had open) that JetBrains IDEs offer that I'm missing here?

**Comment by smbkr** [comment]
The newly released [Sublime Merge](https://www.sublimemerge.com) has a similar implementation.

**Comment by tsachit** [comment]
> This will be a super cool feature, I use `VSCode` for my day job but often resort to `WebStorm` when I need to resolve a merge conflict due to its interactivity and ease to use.

You can add these configuration in your .gitconfig file and type "git mergetool" on your terminal whenever conflict appears. It'll only open your merge gui interface from webstorm for each file.

```
[mergetool "webstorm"]
    cmd = webstorm merge $(cd $(dirname "$LOCAL") && pwd)/$(basename "$LOCAL") $(cd $(dirname "$REMOTE") && pwd)/$(basename "$REMOTE") $(cd $(dirname "$BASE") && pwd)/$(basename "$BASE") $(cd $(dirname "$MERGED") && pwd)/$(basename "$MERGED")
    trustExitCode = true
[merge]
    tool = webstorm
```

**Comment by giordanocardillo** [comment]
+1 here, is now the only reason I use webstorm!

**Comment by krsiakdaniel** [comment]
+1

**Comment by rsimmonsjr** [comment]
+1

**Comment by logusgraphics** [comment]
True, VS Code is amazing except for the lack of this UI tool.

**Comment by liorg2** [comment]
+1, when I have a merge with lots of conflicts, I still open webstorm

**Comment by moonfruit** [comment]
+1, This is exactly what I expect the merge tool to look like.
And #5770 should be implemented like this.

**Comment by michaelKurowski** [comment]
+1 it's the only feature that I'm really missing while using VSC in comparison to WebStorm.

**Comment by danielnice** [comment]
I would very much like this implemented as well, the only big whole I see in VS code as a full IDE

**Comment by doub1ejack** [comment]
Ok, since Webstorm seems to be the only real solution, I'm going to go download that now.  
Really looking forward to VSCode supporting this feature!!

Also, could someone downvote this post for me?  Thanks :P

**Comment by bgarrant** [comment]
Is this feature even in the works?

**Comment by federicotomasi** [comment]
+1, totally would use this if implemented

**Comment by rebornix** [comment]
I added a new setting and command to improve the merge conflict resolution a bit, here is what the experience will be after https://github.com/microsoft/vscode/pull/74231 is merged

Firstly a new setting `merge-conflict.diffViewContext` was added to show the context around the conflict

Before: there is no context around the diff
![compare-withoutcontext](https://user-images.githubusercontent.com/876920/58287661-7c79bd00-7d66-11e9-8a6a-18ab0401ec3c.gif)

After: with `"merge-conflict.diffViewContext": 3`
![compare-with-context](https://user-images.githubusercontent.com/876920/58287490-ffe6de80-7d65-11e9-82c4-fa1fe14d23fd.gif)

Secondly, a new command *Compare All* is introduced to display the full compare between current workspace and incoming changes, with which you can compare the old content, the incoming one, and the latest local content on disk

![compare-all](https://user-images.githubusercontent.com/876920/58287780-bcd93b00-7d66-11e9-8060-f3cc1d55c79c.gif)

More work in the editor area needs to be done to support the JetBrains style, for example, opening a three column editor group from the conflict. But I still want to hear your feedback on the current solution mentioned above.

**Comment by bgarrant** [comment]
This is cool. Can you pick line by line what change you want to merged into the new file or is it just one file or the other? Line by line or All would be awesome!!!

**Comment by bgarrant** [comment]
If you can accept each change from either file block by block and merge into the third that would probably cover most use cases. So for example if one file had 3 changes and the other file had 4 I could chose whatever I wanted. Like accept 2 changes from one file and 2 changes from the other. Can it do this? If so...that rocks!

**Comment by eamodio** [comment]
@rebornix The additional context is nice, and I like the compare all, but for me there still a few things missing:
1. Comparing each "side" with the common base -- while a 3-way merge view is probably best, this could at least help in the interim (I've been toying with adding something for this in GitLens)
1. Being able to edit in the compare view and have the changes sync'd back to the document -- this is something that was talked about before better-merge was brought into the core. Here was a request about it (though the original request is gone :cry:) https://github.com/microsoft/vscode/issues/10547#issuecomment-300201327'

#2 can probably be done today with a custom file system provider (though that feel like a lot of overkill)

**Comment by rebornix** [comment]
@eamodio 

> Comparing each "side" with the common base -- while a 3-way merge view is probably best, this could at least help in the interim (I've been toying with adding something for this in GitLens)

This one requires adding a new three way diff view (a three way diff algorithm and a three column editor) but it's doable. For # 2, I think a custom file system provider is the right way to go as the file content on disk contains merge conflicts (like `>>>>>>, <<<<<<`), while picking content changes from the diff view, we don't want to mess up with the file on disk, so it' a virtual file system provider already.

**Comment by liushiqi9** [comment]
Need this feature very much.

**Comment by andreiucm** [comment]
@rebornix I like the idea I think it can be even better than open a 3 column editor since the laptop screens are not soo big. I think can be easier to improve it a bit to look finally like the Gitkracken merge tool showing the context on the top editor and allowing editing the reluting file (that currently has >>>><<<< ...) on the bottom.
<img width="800" alt="merge-tool@2x" src="https://user-images.githubusercontent.com/19815046/59409429-e19d5d00-8dbe-11e9-9eac-fced6906822d.png">

**Comment by vedmant** [comment]
Yeah, often when there is merge conflict I switch to PHPStorm, even if the other work on the project I do in VSCode. It's weird that two years passed and there is still no appropriate merger.

**Comment by stephanmullerNL** [comment]
> @rebornix I like the idea I think it can be even better than open a 3 column editor since the laptop screens are not soo big. I think can be easier to improve it a bit to look finally like the Gitkracken merge tool showing the context on the top editor and allowing editing the reluting file (that currently has >>>><<<< ...) on the bottom.

I disagree, the three column layout is part of what makes the Webstorm one so great. It keeps a very neat overview

**Comment by fabb** [comment]
Same here, I always switch over to IntelliJ for merge conflict resolution. The „Resolve simple conflicts“ magic wand tool saves so much time too!

**Comment by cailie** [comment]
I change back to Visual Studio 2017 to resolve conflicts...

**Comment by SupinePandora43** [comment]
vscode must change git conflict tool, because while resolving conflicts vscode adds `<<<`, `===`, which causes language syntax error

**Comment by vedmant** [comment]
@SupinePandora43 That's not VSCode, that's how git handles conflicts, it adds `<<<`, `===` to conflicting lines. WebStorm parses this blocks and creates great visual tool for merging.

**Comment by SupinePandora43** [comment]
@vedmant then... vscode needs some parser

**Comment by minkir014** [comment]
See here. Is that what you mean?
https://github.com/Microsoft/vscode-tips-and-tricks/blob/master/media/resolve_merge_conflicts.gif

**Comment by vedmant** [comment]
@minkir014 Check attached image on original comment on top, this what everybody mean.

**Comment by minkir014** [comment]
This is in vs code tips and tricks and that means it's built in vs code. So, why this issue is still opened?

**Comment by jack1142** [comment]
@minkir014 I'll link the image from original comment here again, cause you clearly are thinking about something different: https://user-images.githubusercontent.com/1470309/32250860-c677e4ce-bec0-11e7-82b5-0196d981cc28.png

**Comment by vedmant** [comment]
@minkir014 What's in VS code now is nothing near the feature IntelliJ has, just compare this: https://user-images.githubusercontent.com/1470309/32250860-c677e4ce-bec0-11e7-82b5-0196d981cc28.png and what you posted https://github.com/Microsoft/vscode-tips-and-tricks/blob/master/media/resolve_merge_conflicts.gif do you see much similarities?

**Comment by Arelav** [comment]
@minkir014 because built in feature it's not the same as we are discussing in this thread.

**Comment by jeremy-hughes** [comment]
Three way merge is the only feature blocking me from using VSCode as my only IDE.

**Comment by bgarrant** [comment]
> Three way merge is the only feature blocking me from using VSCode as my only IDE.

I hear ya.  This is a killer feature we need to have.  I have to keep PHPStorm around just for this reason.

**Comment by mscottnelson** [comment]
It took me awhile to realize that you have *most* of what you _need_ contained in updates from the above thread, it's just not displayed pretty, nor can it do arbitrary 3-way diffs (to be super-powered in the future by GitLens with arbitrary commit diffing and ancestor selection, no doubt).

The below config changes cover most real-world merge contextualization needs better than default VSCode/Git settings when you want to be able to understand the context of two different changes when they made while simultaneously reconciling their differences. So, for future searchers:

1. Enable reveal of the common (differing) ancestor of two diffs by changing your git conflict style to diff3. Do this by either editing your ~/.gitconfig or using the cli: `git config merge.conflictStyle diff3`
2. Set your VSCode setting for Diff View Position to "Below" or "Beside" `"merge-conflict.diffViewPosition": "Below"`

Congratulations, you now can see the context of your two diffs as static references, the common base they both branched from, and all while still making edits to all three contexts (ancestor, branch A, and branch B) using in-line diffing. It's now a poor-mans vimdiff.

**Comment by mscottnelson** [comment]
(adding a static pane to display the common ancestor diff without highlights in the middle would be a low-hanging fruit easy addition without requiring editor or diff algo updates, speaking of vimdiff)

**Comment by jeremy-hughes** [comment]
And if these are not changes in GIT? (I'm using perforce, and I don't have a choice in that right now.)

**Comment by Wavez** [comment]
Duplicate of https://github.com/microsoft/vscode/issues/37350

**Comment by AlexAegis** [comment]
> Duplicate of #25887

That's a two column view. This is a three column view.

**Comment by Wavez** [comment]
> > Duplicate of #25887
> 
> That's a two column view. This is a three column view.

I thought it would be worth linking the issues together, cause it's the same problem.
Two or three columns are part of the solution.
On a side note, my motivation for linking is to get attraction from both issues and add the missing behaviour as soon as possible - I would really want to use that feature ☺

**Comment by igolskyi** [comment]
But here are 427 votes against 310 (# 25887). Maybe you can merge count the votes together if you want to merge already. This feature will be number one )

**Comment by oskarspv** [comment]
Implementing an effective and user friendly IntelliJ style 3 pane merge (sometimes refered to as three way merge, but it is two files merged into a result pane) would place the VS Code editor product on the top of the world. I am certain of this - and several of these threads and many comments tell the tale of people "holding on" to their Intellij products just because of the lack of this feature. I don't see how a feature request could ever be better motivated or a clearer cut for the product owners / developers. Any news on your (devs, managers on vs code) thoughts on this?

**Comment by trumbitta** [comment]
This is exactly why I **still** use Meld as my mergetool

**Comment by tsachit** [comment]
> This is exactly why I **still** use Meld as my mergetool

Sublime merge is far better than meld

**Comment by jack1142** [comment]
> > This is exactly why I **still** use Meld as my mergetool
> 
> Sublime merge is far better than meld

I agree, though I don't think this is the place to discuss what merge tool is the best one out there.

**Comment by ackvf** [comment]
Is it at least possible to show a common BASE for both Current and Incomming changes?

**Comment by blacksteed232** [comment]
@ackvf yes, you can with [three-way diff](https://stackoverflow.com/questions/18131515/how-can-i-see-a-three-way-diff-for-a-git-merge-conflict)

**Comment by anrananran** [comment]
so,  What's new?

**Comment by jqweryjson** [comment]
Waitng too, anybody plz notice me when its available

**Comment by mcianc** [comment]
waiting the same...have that tool will be amazing

**Comment by namtran1501** [comment]
maybe i should get IntelliJ too :)

**Comment by Arelav** [comment]
> +1 here, is now the only reason I use webstorm!
> +1
> waiting the same...have that tool will be amazing
> maybe i should get IntelliJ too :)

Please stop spamming all subscribers of this ticket with non-related messages.

**Comment by carlespibernat** [comment]
I have PyCharm CE (which is free) and I just use it for my PHP and javascript merges

**Comment by jklassen** [comment]
Has VS Code added this feature? This is closed...

**Comment by sangnguyen7** [comment]
I think this extension (https://github.com/mhutchie/vscode-git-graph) almost has alike features as in Webstorm IDE.

**Comment by mbrookson** [comment]
This is the main reason I don't use vscode as my main editor. Would be great to see this added as the built-in git integration and merging is so much more helpful in WebStorm. Would be great to see this added.

**Comment by abernier** [comment]
> This is the main reason I don't use vscode as my main editor. Would be great to see this added as the built-in git integration and merging is so much more helpful in WebStorm. Would be great to see this added.

@mbrookson you finished to convince me to finally quit this trolling issue => unsubscribing 👋🏻

**Comment by arielalvarez88** [comment]
I need this in my life.

**Comment by fredconv** [comment]
No news about this feature ? or even a plugin is fine :D  ?? 

vscode is great but indeed still using intellij merge tool just for that .. )

**Comment by jack1142** [comment]
All the news we have is the 2020 Roadmap which mentions "full merge support (3-way)": https://github.com/microsoft/vscode/wiki/Roadmap#scm

**Comment by hhimanshu** [comment]
Oh man! such a big miss from VSCode. I wonder how developers survive without this feature. I am going to resort to Webstorm for this

**Comment by Baindaer** [comment]
Checked the Git Graph extension for Visual Studio Code but still isn't the feature. Im using vscode for all but merge versions selective with pycharm.

**Comment by CatalanCabbage** [comment]
_Still_ using IntelliJ for merging 😐

**Comment by rxdeveloper** [comment]
Man I created this request 3 years ago :)

**Comment by piavgh** [comment]
Though VSCode is great, I still have to use Webstorm for Git related tasks :(

**Comment by g707175425** [comment]
I'm waiting for it

**Comment by sshsu** [comment]
why still no support this? is there any vscode plugin play the same role in the market?

**Comment by pagdot** [comment]
@sshsu 
I may be wrong, but I think the limiting factors are missing editor features. If it would be possible to write a plugin, there would already be one

**Comment by shaman-apprentice** [comment]
> I may be wrong, but I think the limiting factors are missing editor features. If it would be possible to write a plugin, there would already be one

Kind of. I started to implement a merge plugin in my free time and the only thing I found to be really missing was https://github.com/microsoft/vscode/issues/85682 

However, I am really grateful for the open source ide vscode 😍 So I will just be happy about everything what will be added - maybe we will even see this feature this year, as it is part [of the road map 2020](https://github.com/microsoft/vscode/wiki/Roadmap#scm) 😍😍

**Comment by Cloudmersive** [comment]
Is this feature ever coming?  Visual studio mainline has 3-way merge in TFS

**Comment by zawys** [comment]
You maybe want to try out my new extension [VS Code as Git Mergetool](https://github.com/zawys/vscode-as-git-mergetool). It has some deficits due to lacking API but I have not received any negative feedback yet, except from myself. The proposed four-pane layout could even have some advantages over a three-column layout, however.

It’s not yet in the Marketplace, but it can be installed manually.

**Comment by abdulrahman-khankan** [comment]
Until this feature comes out ([as mentioned in the 2020 roadmap](https://github.com/microsoft/vscode/wiki/Roadmap#scm)), if you are working on a non-commercial project, you can use [Gitkraken for free](https://www.gitkraken.com/invite/bunztBRx), it has a really nice 3-way merge tool.

Looking forward to see this in VSCode!

**Comment by nomi9995** [comment]
I will delete webstorm when I get this feature.

**Comment by jcbjorgensen** [comment]
Much needed :+1:

**Comment by razvantomareea** [comment]
This will be a great feature <3

**Comment by FDiskas** [comment]
As a temporary solution is use online tools like https://editor.mergely.com

**Comment by AlexAegis** [comment]
I have a comment here from a few years ago saying the same thing as others, and while I think an alternative three way merge view would be awesome, I actually prefer this barebones way now since it's literally just a skin over the text that git puts there.

What I want to say that this should not be a "blocker" for anyone using VSCode.

**Comment by jcimoch** [comment]
What is the status on this?

**Comment by FDiskas** [comment]
Actually it could be done just by removing read only flag on comparing both changes page :)

**Comment by AyushiBShah** [comment]
@FDiskas : Can you please elaborate ?

**Comment by FDiskas** [comment]
Currently if conflicted file is opened and you can click on "Compare changes" and it opens side by side and thous are readonly. Make them writable same as https://editor.mergely.com

**Comment by rjmunro** [comment]
This proposed view does not really help you resolve the conflict. What you actually need to see is the common parent of the 2 conflicting changes and what each of the diffs are - i.e. the information you get from `git config merge.conflictstyle diff3`. See https://blog.nilbus.com/take-the-pain-out-of-git-conflict-resolution-use-diff3/ for examples of why this is useful.

Could this be taken into account in any kind of multi-panel diff type conflict resolution view?

**Comment by jcimoch** [comment]
I see there is https://github.com/microsoft/vscode/wiki/Roadmap "provide 3 merge support" , however I don't see roadmap for 2021, is it still planned?

**Comment by sayjeyhi** [comment]
Not only to merge conflicts but I also need this to have a better experience on committing my local changes. (InellijIdea has really better experience)

**Comment by Borda** [comment]
Any update here? seems to be a hot topic for more than 4years now...
...and for me, as a git user cooperation in a team, this is a critical feature 🐰

**Comment by rebosante** [comment]
I'm still waiting for this feature... will I die before it comes to life?

**Comment by Ramonkos** [comment]
Tik tok tik tok....
Still waiting!

**Comment by rebosante** [comment]
> I have made a `vscode extension` for the open IntelliJ tool for merging from vs code
> 
> extension name: `IntelliJ Merge Tool
> 
> github: https://github.com/nomi9995/intellij-merge-tool
> 
> # usage
> you can right mouse click on the conflicted file and click on `merge with intellj`
> 
> <img alt="" width="30%" src="https://raw.githubusercontent.com/nomi9995/intellij-merge-tool/HEAD/media/example1.png">
> 
> ![image](https://user-images.githubusercontent.com/36044436/111220052-83885400-85fa-11eb-8ba0-ec263a359203.png)
> 
> you can check it out

Hi @nomi9995, In order to use your extension..  Do I need any intellij software installed on my computer anyway?
Is it required?

**Comment by lolmaus** [comment]
@rebosante Yes. This extension simply opens an IntelliJ IDE to resolve the conflict.

**Comment by FDiskas** [comment]
@nomi9995 can you integrate something like http://meldmerge.org/ or https://www.perforce.com/products/helix-core-apps/merge-diff-tool-p4merge or other free tool?

**Comment by matiusmonkey** [comment]
Can I also add a yes please to this feature

**Comment by carwyn** [comment]
Related: #5770 and #25887

**Comment by jochman** [comment]
+1 - It will be awesome!

**Comment by trumbitta** [comment]
@swtch1 so you are telling us Microsoft is not paying its own developers to work on VSCode during their usual work hours as a part of its OS strategy? Well, if that's the (unlikely) case, they're doing it wrong.

This repository is literally under the "microsoft" organization on GitHub.

To be clear, we are paying for this software with our ever increasing fidelity to the Microsoft brand.

On the topic at hand: three-way merging is just how you do merging on a GUI. And it's not an IntelliJ thing: there are several other softwares out there that do it. Because that's how you do it on a GUI.
Anything short of that is a workaround I would expect from a startup, not Microsoft.

**Comment by swtch1** [comment]
@trumbitta I'm confident Microsoft pays devs to work on this, and I understand it's not a 100% charitable act.  What I'm saying is that it makes me sad to see issues boards of open source projects turn into the comments section on YouTube.  Where mostly anonymous people say whatever makes them feel better in that instant without concern for the people on the other side.  That's it.  Everyone has a message to get across, but there's rarely reason to be an asshole about it.

And yeah, there are plenty of diff tools you can use to get a solid comparison and merge interface right now, right inside vscode.

**Comment by Borda** [comment]
> What I'm saying is that it makes me sad to see issues boards of open source projects turn into the comments section on YouTube.

Hi there, I believe I can say I am an open-source dev, so I know the case from both sides of the "barricade"...
I would say that the main frustration comes from the long years it is hanging (it is unresolved for almost 4 years), so one possible conclusion would be to close it and clearly say that it won't be implemented :rabbit:

**Comment by jcimoch** [comment]
I think people are getting rude and furious because it's 
a) Open from 2017
b) it was on a 2020/21 roadmap
c) it got deleted from the roadmap
d) the issue is still open without clear communication from MS despite high demand across years for a basic feature

**Comment by nomi9995** [comment]
hi, @bpasero @jrieken @joaomoreno @mjbvz top contributors of vs code when it will add in vs code?. if `Microsoft` can pay for this then I can manage this by open source developers team to see this issue

**Comment by laurentlb** [comment]
Hi,

Would you accept PRs for this feature?
Who should I sync with before starting the design?

**Comment by Zielak** [comment]
Instead of direct PRs here you could try making an extension?

**Comment by nomi9995** [comment]
@laurentlb I think we should make a `slack` channel for it and add other opensource developers to it. I will be the first. for design and functionality, we will make the same as it is `IntelliJ, WebStorm`

**Comment by jochman** [comment]
> Instead of direct PRs here you could try making an extension?

I'm supporting it as a temporary solution as it much faster to release extensions.

**Comment by PeterChen1997** [comment]
mark

**Comment by apgapg** [comment]
The day this issue is resolved i will switch back to vscode from intelliJ. The biggest pro of IntelliJ is refactor and VCS.

**Comment by niccolomineo** [comment]
We are all waiting for this to become reality...

**Comment by LiranBri** [comment]
+1

**Comment by appleneilAM** [comment]
+1

**Comment by gabdsg** [comment]
+1

**Comment by jvanderen1** [comment]
Can we lock comments until someone at Microsoft gives an update?

**Comment by FDiskas** [comment]
There is not bad project
https://github.com/moshfeu/vscode-diff-merge
Sure it is doing different things, but author likes the idea about extending functionality to include git conflicts solving
Anybody is welcome to collaborate

Offtopic:
I don't think that comments should be locked.
I also don't like thous +1 comments for just subscribing this

**Comment by fbrzlarosa** [comment]
+1

**Comment by DaiQiangReal** [comment]
> that's the main reason I still use Webstorm at my workplace
that's the all reason I still use Webstorm at my workplace

**Comment by eacousineau** [comment]
For people wondering about extensions, "VS Code as Git Mergetool" seems promising! (3-way and 4-way diff layouts)
https://github.com/zawys/vscode-as-git-mergetool

**Comment by michaelblyons** [comment]
It'd be nice if a repo admin cleaned up all the comments that do not add value.
- "Here's a related plugin": keep.
- "My dog wants this feature, too": hide/delete.

**Comment by jmevel** [comment]
> It'd be nice if a repo admin cleaned up all the comments that do not add value.
> 
> * "Here's a related plugin": keep.
> * "My dog wants this feature, too": hide/delete.

I understand you find some comments polluting this thread however if you take a look at the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct) page you'll see those comments are not listed as `unacceptable behavior`.

We've so many of these messages in this thread due to the facts that
- This is a 4 years old request
- This is a very popular request with 1308 👍 as of today ([in fact this is the 3rd most wanted feature](https://github.com/microsoft/vscode/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc))

So to sum-up, yes this kind of messages will continue to popup, yes we will continue to receive notifications for each one of them but we just have to bear with them until Microsoft or someone else implements it.

Now back on the main topic

> For people wondering about extensions / plugins, "VS Code as Git Mergetool" seems promising! (3-way and 4-way diff layouts) https://github.com/zawys/vscode-as-git-mergetool

Thanks a lot for this link, this repo looks interesting, I'm going to give it a try some time

**Comment by bugchecker** [comment]
When you add 3-way merge comparison, don't forget to add a hot switch between 3-way and 2-way (without application reloading or config editing)

**Comment by tjx666** [comment]
Seems cool!

**Comment by trumbitta** [comment]
@gaudibr it would be of extreme value to me especially now that I'm using Gitpod (but I think the same applies to GitHub Codespaces?) all the time, for everything

**Comment by gaudibr** [comment]
@trumbitta I think it would be of great value to everybody using vscode remote and I'm currently struggling with more complex merge conflicts that can't be resolved with a simple "accept incoming"/"accept current". Although there's a neat "compare both" options in the conflict viewer, it baffles me that the compare editor is still read only. I wonder if there's a way to reopen that suggestion or revisit the validation process that decided not to go for it.

**Comment by Zerotask** [comment]
In Visual Studio this is so much better than in Visual Studio Code.
I noticed that beginners of my company / friends had trouble to understand what is current and what is incoming.

**Comment by andreribeiro27** [comment]
This would be amazing, the current VS Code way is much more prone to errors than Webstorm.

**Comment by Igorgro** [comment]
This issue was included in the next Iteration #144519 🎉

**Comment by michaelblyons** [comment]
>  🔴 Explore 3-way merge editor

"Explore" sounds like a good place to start, at least.

**Comment by SidBlack** [comment]
Hello, I have a question, I do not know why you are asking me with this
problem I do not know programming I do not know what you mean, this is not
my field

pon., 7 mar 2022 o 00:03 Michael ***@***.***> napisał(a):

**Comment by tarekahf** [comment]
This is planned for release... That's awesome!

**Comment by linqunhe** [comment]
wow, nice feature~ , i like it

**Comment by nomi9995** [comment]
According to the plan https://github.com/microsoft/vscode/issues/144519, it will ship in **early April**.

**Comment by g707175425** [comment]
666
  
​

**Comment by elazarcoh** [comment]
that task says "explore", I'm not sure what that means... 

I'm just saying it might not be April just yet.

I'm just hoping there'll be an API that'll allow us to implement different diff algorithms (as there are multiple of those, some are better than the others).

**Comment by passmesomesugar** [comment]
Its march 2022 lol. Is it still not released?

**Comment by nomi9995** [comment]
@daviddossett @isidorn what is the status of 3 way merge because, in the iteration plan, you are working 3-way merge.

**Comment by isidorn** [comment]
We are working on 3-way merge - we are in the investigation phase. As soon as we have updates we will share here. We hope this is something that will happen this year.

**Comment by cloudhan** [comment]
This calendar year or this fiscal year? If it is fiscal year, then it should be landed before July!

**Comment by sb-baris** [comment]
> This calendar year or this fiscal year? If it is fiscal year, then it should be landed before July!

Fiscal year for which country? or even for which company?

**Comment by cloudhan** [comment]
> Fiscal year for which country? or even for which company?

FYI, July 1st is the new year for Microsoft.

**Comment by nemchik** [comment]
Hassling contributors for ETAs isn't going to get it released faster. Please be appreciative that someone has taken the time to put their efforts into making this happen for us all.

**Comment by machinecode** [comment]
Hi @isidorn, my guess is most of the people here expect a 3-way merge screen with a magic wand like Intellij had. Please consider having such that feature too.

**Comment by hediet** [comment]
In the previous release we shipped the basic merge editor ([release notes](https://code.visualstudio.com/updates/v1_69#_3-way-merge-editor)). In the release next week, we are going to enable it by default.
Also, the next release includes CLI arguments to configure VS Code as merge tool.

Please let us know what you think about the new merge editor, so we can continue improving it!
